st3/lsp_utils/_client_handler/language_handler.py

```diff
@@ -0,0 +1,146 @@
+from LSP.plugin import ClientConfig
+from LSP.plugin import LanguageHandler
+from LSP.plugin import Notification
+from LSP.plugin import read_client_config
+from LSP.plugin import Request
+from LSP.plugin import Response
+from LSP.plugin.core.typing import Any, Callable, Dict, Optional
+from sublime_lib import ActivityIndicator
+import sublime
+
+__all__ = ['ClientHandler']
+
+
+class ApiWrapper(ApiWrapperInterface):
+    def __init__(self, client):
+        self.__client = client
+
+    # --- ApiWrapperInterface -----------------------------------------------------------------------------------------
+
+    def on_notification(self, method: str, handler: Callable[[Any], None]) -> None:
+        self.__client.on_notification(method, handler)
+
+    def on_request(self, method: str, handler: Callable[[Any, Callable[[Any], None]], None]) -> None:
+        def on_response(params, request_id):
+            handler(params, lambda result: send_response(request_id, result))
+
+        def send_response(request_id, result):
+            self.__client.send_response(Response(request_id, result))
+
+        self.__client.on_request(method, on_response)
+
+    def send_notification(self, method: str, params: Any) -> None:
+        self.__client.send_notification(Notification(method, params))
+
+    def send_request(self, method: str, params: Any, handler: Callable[[Any, bool], None]) -> None:
+        self.__client.send_request(
+            Request(method, params), lambda result: handler(result, False), lambda result: handler(result, True))
+
+
+class ClientHandler(LanguageHandler, ClientHandlerInterface):
+    _setup_called = False
+
+    # --- LanguageHandler handlers ------------------------------------------------------------------------------------
+
+    @property
+    def name(self) -> str:
+        return self.get_displayed_name().lower()
+
+    @classmethod
+    def additional_variables(cls) -> Optional[Dict[str, str]]:
+        return cls.get_additional_variables()
+
+    @property
+    def config(self) -> ClientConfig:
+        settings, filepath = self.read_settings()
+        settings_dict = {}
+        for key, default in self.get_default_settings_schema().items():
+            settings_dict[key] = settings.get(key, default)
+        if self.manages_server():
+            can_enable = self.get_server() is not None
+        else:
+            can_enable = True
+        enabled = settings_dict.get('enabled', True) and can_enable
+        settings_dict['enabled'] = enabled
+        if not settings_dict['command']:
+            settings_dict['command'] = self.get_command()
+        return read_client_config(self.name, settings_dict, filepath)
+
+    @classmethod
+    def on_start(cls, window: sublime.Window) -> bool:
+        if cls.manages_server():
+            server = cls.get_server()
+            return server is not None and server.get_status() == ServerStatus.READY
+        message = cls.is_allowed_to_start(window)
```

Comment:
Ugh, so the old LanguageHandler doesn't receive the ClientConfig?

Reply:
Maybe we could store the ClientConfig that we are creating in `config()` and then we could pass it here. Just not sure whether that could end up with the one that LSP has and the one that LSP-* has, getting out of sync (if LSP would re-create its own).

Reply:
LSP always creates a copy in WindowConfigManager.update

Reply:
Then I guess we would have to add something on the master branch for that? But the old API was historically hard to extend so not sure how that would work out.